support KHR_xmp_json_ld read in cgltf.h

PsycoTodd opened the pull request:
Hello, 

This is a change that try to support [KHR_xmp_json_ld extensions](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_xmp_json_ld/README.md).

Basically, the extension allow to provide json_ld format things as array of packets at top level, and refer by index in `asset, scene, node, mesh, material, image, animation`. 

The current read implementation is for expose the data at each component as a new array of xmp_json_ld, since it is possible to refer multiple packets by one, for example, image/node, etc.

Will see the feedback and try to also implement the cgltf_write.h functions.

Test pass by locally create model with the field and load inside filament.

**One small change:**

I notice that we create extensions for each component, and allocate array with total extension size, but only feel the array when we encounter unprocessed extensions. Or other recognizable extensions are processed in other if branch. This seems to make a extension array potentially larger than it needs. Since it only hold unprocessed extensions. 

I made a small change to get the number of valid/supported extensions first, then allocate the extensions array accordingly.

zeux commented:
On the data structure, I feel like using `cgltf_xmp_json_packet** xmp_packets; size_t xmp_packets_count;` instead of a struct with a packet index would fit the rest of the API better. We can add a cgltf_xmp_packet_index similar to all other _index functions to go back from a pointer to an index. You already see this pattern eg with child nodes for cgltf_node.

On unprocessed extensions, I would strongly recommend moving this out of this PR. This is a problem worth attacking but it should not be part of a PR that adds a new extension, and it should be done in a way that is minimaly invasive on existing code - eg repeating extension names will result in subtle bugs in the future, so we need to find a different way to do this, for example lazily reallocating unprocessed extensions as we encounter them.

On actual XMP parsing code, I think this needs to be less repetitive. Ideally every object that needs XMP (thank you for restricting these to the top level important objects! that's a good call) should just need a couple extra lines to process these. It seems like this is missing some refactoring on the parser side.

PsycoTodd commented:
**On actual XMP parsing code, I think this needs to be less repetitive. Ideally every object that needs XMP (thank you for restricting these to the top level important objects! that's a good call) should just need a couple extra lines to process these. It seems like this is missing some refactoring on the parser side.**

I think in the same way let me think a little. But I just notice that different compoenent may have different extension parsing logic. Like materials has a lot of extensions to parse to its own structure, while others may not. What is the method in C we recommend to have this behavior difference? To me seems like for node, mesh, image, etc. the way to parse the extensions list, the process is similar, but the detail would differ due to extension would be different. 

This seems make the code has to repeat somehow on the structure level.

PsycoTodd commented:
Updated with the comments.

* Removed the extension count functions.
*  add only code to handle xmp cases.
* Use pointer reference rather than the index to trace the json string.
* Each component can only have one KHR_xmp_json_ld index, so we do not use array anymore.

zeux commented on cgltf.h:
```diff
@@ -3338,7 +3372,36 @@ static int cgltf_parse_json_mesh(cgltf_options* options, jsmntok_t const* tokens
+
+			int extensions_size = tokens[i].size;
+			++i;
+			out_mesh->extensions_count = 0;
```
This is missing extensions allocation

zeux commented on cgltf.h:
```diff
@@ -4269,7 +4332,35 @@ static int cgltf_parse_json_image(cgltf_options* options, jsmntok_t const* token
+				return CGLTF_ERROR_JSON;
+			}
+
+			int extensions_size = tokens[i].size;
```
Missing extensions allocation and extensions_count setup

zeux commented on cgltf.h:
```diff
@@ -4269,7 +4332,35 @@ static int cgltf_parse_json_image(cgltf_options* options, jsmntok_t const* token
+			int extensions_size = tokens[i].size;
+			++i;
+
+			for (int j = 0; j < extensions_size; ++j) {
```
```suggestion
			for (int j = 0; j < extensions_size; ++j)
			{
```

zeux commented on cgltf.h:
```diff
@@ -6053,6 +6238,42 @@ static int cgltf_parse_json_asset(cgltf_options* options, jsmntok_t const* token
+
+static int cgltf_parse_xmp_json_ld_packets(cgltf_options* options, jsmntok_t const* tokens, int i, const uint8_t* json_chunk, cgltf_data* out_data) {
+
+	i = cgltf_parse_json_array(options, tokens, i, json_chunk, sizeof(cgltf_material_variant),
```
The sizeof is incorrect

zeux commented on cgltf.h:
```diff
@@ -2038,6 +2055,8 @@ void cgltf_free(cgltf_data* data)
 
 	data->memory.free_func(data->memory.user_data, data->variants);
 
+	data->memory.free_func(data->memory.user_data, data->xmp_json_ld_packets);
```
This needs to free contents of each packet

zeux commented on cgltf.h:
```diff
@@ -6267,6 +6488,29 @@ static int cgltf_parse_json_root(cgltf_options* options, jsmntok_t const* tokens
+					for (int m = 0; m < data_size; ++m) {
+						CGLTF_CHECK_KEY(tokens[i]);
+
+						if (cgltf_json_strcmp(tokens + i, json_chunk, "packets") == 0) {
```
Can this code just call cgltf_parse_xmp_json_lds?

zeux replied:
Oh nvm this is the actual packet parsing

zeux commented on cgltf.h:
```diff
@@ -6267,6 +6488,29 @@ static int cgltf_parse_json_root(cgltf_options* options, jsmntok_t const* tokens
+
+						if (cgltf_json_strcmp(tokens + i, json_chunk, "packets") == 0) {
+							i = cgltf_parse_xmp_json_ld_packets(options, tokens, i + 1, json_chunk, out_data);
+						} else {
```
Please fix bracing style here and everywhere else in the patch to conform to the style - braces should be on their own line.

PsycoTodd replied:
sure. will do.

zeux commented on cgltf.h:
```diff
@@ -5912,7 +6041,35 @@ static int cgltf_parse_json_animation(cgltf_options* options, jsmntok_t const* t
+
+			int extensions_size = tokens[i].size;
+			++i;
+			out_animation->extensions_count = 0;
```
Missing allocation

PsycoTodd commented:
Hello, please take a look with the new cr. thanks.
